pkg/csplugin/broker.go

```diff
@@ -0,0 +1,339 @@
+package csplugin
+
+import (
+	"context"
+	"fmt"
+	"io/fs"
+	"os"
+	"os/exec"
+	"os/user"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall"
+	"text/template"
+	"time"
+
+	"github.com/crowdsecurity/crowdsec/pkg/csconfig"
+	"github.com/crowdsecurity/crowdsec/pkg/models"
+	"github.com/crowdsecurity/crowdsec/pkg/protobufs"
+	plugin "github.com/hashicorp/go-plugin"
+	log "github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	PluginProtocolVersion uint   = 1
+	CrowdsecPluginKey     string = "CROWDSEC_PLUGIN_KEY"
+)
+
+type PluginBroker struct {
+	PluginChannel                   chan ProfileAlert
+	alertsByPluginName              map[string][]*models.Alert
+	profileConfigs                  []*csconfig.ProfileCfg
+	pluginConfigByName              map[string]PluginConfig
+	pluginMap                       map[string]plugin.Plugin
+	notificationConfigsByPluginType map[string][][]byte // "slack" -> []{config1, config2}
+	notificationPluginByName        map[string]Notifier
+	watcher                         PluginWatcher
+	pluginKillMethods               []func()
+}
+
+// holder to determine where to dispatch config and how to format messages
+type PluginConfig struct {
+	Type           string        `yaml:"type"`
+	Name           string        `yaml:"name"`
+	GroupWait      time.Duration `yaml:"group_wait"`
+	GroupThreshold int           `yaml:"group_threshold"`
+	MaxRetry       int           `yaml:"max_retry"`
+	TimeOut        time.Duration `yaml:"timeout"`
+
+	Format string `yaml:"format"` // specific to notification plugins
+
+}
+
+type ProfileAlert struct {
+	ProfileID uint
+	Alert     *models.Alert
+}
+
+func (pb *PluginBroker) Init(profileConfigs []*csconfig.ProfileCfg, configPaths *csconfig.ConfigurationPaths) error {
+	pb.PluginChannel = make(chan ProfileAlert)
+	pb.notificationConfigsByPluginType = make(map[string][][]byte)
+	pb.notificationPluginByName = make(map[string]Notifier)
+	pb.pluginMap = make(map[string]plugin.Plugin)
+	pb.pluginConfigByName = make(map[string]PluginConfig)
+	pb.alertsByPluginName = make(map[string][]*models.Alert)
+	pb.profileConfigs = profileConfigs
+	if err := pb.loadConfig(configPaths.NotificationDir); err != nil {
+		return err
+	}
+	err := pb.loadPlugins(configPaths.PluginDir)
```

Comment:
Maybe check if err != nil, and returns directly without trying to init the watcher ?

Reply:
thanks ! done via https://github.com/crowdsecurity/crowdsec/pull/857/commits/bf954ccd25869f1a631a588177d18f3ccd14bc38